### PR TITLE
module 3 tests and refactoring

### PR DESCRIPTION
**Pull request by MHolmes219**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,3 +57,16 @@ You should see output similar to the following:
  * Debugger PIN: 226-556-590
 ```
 Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser to view the app.
+
+## Running tests
+
+In order to run tests, you will need to install pytest using `pip install pytest`. You can then run all of your tests in your terminal, with optional verbosity to check output:
```
Comment (JackMeadDev) [inline README.md]: We don't want to install `pytest` in this fashion, since we want to avoid users having to add all their own packages and potentially clashing versions; this is one advantage of our virtual environment. We should instead register the dependency with `poetry add pytest` and future users will automatically pick it up with `poetry install`
(Alternatively, run `poetry add pytest -D` to register it as a dev-dependency, so we can easily exclude it from production builds)

File: tests/test_app.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import os
```
Comment (JackMeadDev) [inline tests/test_app.py:1]: It's not necessary here, but adding a `__init__.py` to this folder to acknowledge it as a Python package may save you from pain in later modules - particularly when we're using Docker. 
When running pytest outside my virtual environment (as your instructions hint), I saw issues due to invalid module names, which required me to add that file to fix; another advantage of our virtual environment making things more consistent!

File: tests/test_app.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        }]
+    return StubResponse(fake_response_data)
+
+
+def test_index_page(monkeypatch, client):
+    monkeypatch.setattr(requests, 'get', get_lists_stub)
+    response = client.get('/')
+
+    assert response.status_code == 200
+    assert 'Test card' in response.data.decode()
```
Comment (JackMeadDev) [inline tests/test_app.py:45]: Nice; clean, simple and testing effectively.

File: tests/test_view_model.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    view_model = ViewModel(items)
+
+    assert len(view_model.to_do_items) == 0
+    assert len(view_model.in_progress_items) == 0
+    assert len(view_model.done_items) == 0
+
+
+def test_items_are_split_into_categories():
+    items = [
+        Item(1, 'Task 1', today, '', "2022-02-08T17:36:12.1345Z", 'Something'),
```
Comment (JackMeadDev) [inline tests/test_view_model.py]: Nice choice here to throw in some valid but irrelevant data to check it can handle the case cleanly

File: tests/test_view_model.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
+from datetime import datetime, timedelta
+from todo_app.data.trello_items import Item
+from todo_app.view_model import ViewModel
+
+yesterday = datetime.now() - timedelta(1)
+yesterday = yesterday.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+today = datetime.now().strftime("%Y-%m-%dT%H:%M:%S.%fZ")
```
Comment (JackMeadDev) [inline tests/test_view_model.py]: This is a nice solution the problem, but it still could be hit by flaky behaviour due to changing times - e.g. maybe the code works most days but not on February 29th in leap years!
This changing behaviour can have a helpful side effect for flushing that out, but generally we want to keep our tests consistent, and have explicit tests for those edge cases if we're concerned, so we aim to "freeze" time so that the tests always believe it to be a fixed time. You could take a look at [pytest-freezegun](https://pypi.org/project/pytest-freezegun/) as one example of a library that can help with achieving that

File: todo_app/templates/index.html [hunk 6]
```diff
@@ -69,6 +79,22 @@ <h6 class="card-subtitle font-weight-normal text-muted mt-2">{{ card.description
       </form>
     </div>
   </div>
+
+  {% if request.endpoint == 'index' %}
```
Comment (JackMeadDev) [inline todo_app/templates/index.html]: This makes me a little nervous - it suggests this file is trying to handle two cases (breaking our single-responsibility-principle); rather than having the index & view_card endpoints sharing the same HTML, can we extract the shared sections into layouts, and commonise them that way?

File: todo_app/view_model.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+class ViewModel:
+    def __init__(self, items):
+        self._items = items
+
+    @property
+    def to_do_items(self):
```
Comment (JackMeadDev) [inline todo_app/view_model.py:6]: All really nice; simple and clear

File: todo_app/templates/index.html [hunk 8]
```diff
@@ -1,4 +1,61 @@
...
+    <button type="submit" class="btn btn-sm btn-primary" data-toggle="tooltip" data-placement="right" title="Mark as complete?">Complete</button>
+  </form>
+  {% endif %}
+  <form id="archive-card" method="post" action="/archive-card/{{ item.id }}">
+    <button type="submit" class="btn btn-link btn-lg" data-toggle="tooltip" data-placement="right" title="Archive?"><i class="bi bi-x-circle"></i></button>
+  </form>
+</li>
+{%- endmacro %}
+
+{% macro show_item_list(title, items, section_id) -%}
```
Comment (JackMeadDev) [inline todo_app/templates/index.html:40]: Nice use of these macros

File: todo_app/app.py [hunk 9]
```diff
@@ -1,69 +1,92 @@
...
+def suffix(d):
+    return 'th' if 11<=d<=13 else {1:'st',2:'nd',3:'rd'}.get(d%10, 'th')
+
+def custom_strftime(format, t):
+    return t.strftime(format).replace('{S}', str(t.day) + suffix(t.day))
+
+@filters.app_template_filter()
+def formatted_time(dttm):
+    ettm = datetime.datetime.strptime(dttm, "%Y-%m-%dT%H:%M:%S.%fZ")
+    return custom_strftime("{S} %B %Y", ettm)
```
Comment (JackMeadDev) [inline todo_app/app.py:92]: This is a nice solution I haven't seen before, but a simpler one might involve registering this as a function on your `Item` class - that has the advantage of being less Flask specific, and of letting you call the formatted time anywhere you like, and not just within the templates

**Review by JackMeadDev — APPROVED** [review]
Nice work, I was able to run your app & tests easily, and your site looks great! The only real issue is that your tests didn't work out of the box because you're not relying on a tracked `pytest` installation in your virtual environment so see my comment below about installing that and running the tests. Otherwise I've left some optional suggestions below, but I'm happy for you to merge this when you're ready.
